README.md
TYPO3 CMS Extension: Nagios
===========================

Nagios TYPO3 CMS Monitoring (TYPO3 extension "nagios")

Description
-----------

Nagios is the industry standard in open source server/service/network monitoring.
This TYPO3 CMS extension gathers information about the TYPO3 instance for a Nagios monitoring server.

Read full documentation at: https://schams.net/nagios

Find out more about TYPO3 CMS at: http://typo3.org

License
-------

(c) 2010-2016 Michael Schams (schams.net), all rights reserved

This software free software; you can redistribute it and/or modify it under the terms of the GNU General Public License, either version 2 of the License, or any later version. For the full copyright and license information, please see the documentation that was distributed with this source code.

The GNU General Public License can be found at:
http://www.gnu.org/copyleft/gpl.html


Author
------
Michael Schams - https://schams.net


ChangeLog
2016-04-09 Michael Schams <schams.net>

	* documentation and ChangeLog updated
	* version raised to 1.2.12
	* extension released as version 1.2.12 in TYPO3 Extension Repository (TER)

2016-01-31 Michael Schams <schams.net>

	* allow for multiple IP addresses sent by proxies/caches/load balancers

2015-08-31 Michael Schams <schams.net>

	* allow hostnames in list of allowed Nagios servers (merged)

2015-08-10 Gernot Leitgab <leitgab@gmail.com>

	* allow hostnames in list of allowed Nagios servers (pull-request)

2014-08-24 Michael Schams <schams.net>

	* documentation updated (typos, styles and formatting corrected)
	* code clean up to comply with SonarSource CGL: https://metrics.typo3.org/dashboard/index/org.typo3:extension-nagios
	* Transfer source code to GitHub: https://github.com/schams-net/nagios

2014-07-11 Michael Schams <schams.net>

	* documentation updated and converted into ReST format
	* LICENSE.txt file added (GPLv2)
	* header comments simplified/updated
	* deprecation warning eliminated: EidUtility::connectDB()
	* extension released as version 1.2.11 in TYPO3 Extension Repository (TER)

2014-03-22 Michael Schams <schams.net>

	* documentation updated
	* extension released as version 1.2.10 in TYPO3 Extension Repository (TER)

2014-01-31 Michael Schams <schams.net>

	* internal tests completed and released to beta testers
	
2013-12-13 Michael Schams <schams.net>

	* support for generating extension list (loaded extensions only) for TYPO3 CMS 6.2 added
	* deprecated code to use extension as a (frontend) plugin removed
	* language files simplified and combined
	* extension in beta test

2013-05-31 Michael Schams <schams.net>

	* tested with TYPO3 CMS 6.1 and compatibility statement updated
	* documentation updated
	* extension released as version 1.2.9 in TYPO3 Extension Repository (TER)

2013-03-13 Michael Schams <schams.net>

	* (bugfix) PHP warnings in sys log ("in_array() expects parameter 2 to be array [...] t3lib/class.t3lib_iconworks.php")
	* extension released as version 1.2.8 in TYPO3 Extension Repository (TER)

2013-01-09 Michael Schams <schams.net>

	* access verification when checking for HTTP proxy headers fixed
	* documentation updated
	* extension released as version 1.2.7 in TYPO3 Extension Repository (TER)

2012-12-17 Michael Schams <schams.net>

	* documentation updated
	* extension released as version 1.2.6 in TYPO3 Extension Repository (TER)

2012-12-15 Michael Schams <schams.net>

	* usage of the extension as a FE plugin discontinued and partly removed

2012-12-11 Michael Schams <schams.net>

	* take IP addresses into account, which are typically passed through in HTTP headers from a proxy/cache/load balancer

2012-11-06 Michael Schams <schams.net>

	* documentation updated
	* extension released as version 1.2.5 in TYPO3 Extension Repository (TER)

2012-11-02 Michael Schams <schams.net>

	* imported but not loaded extensions included in extension list (requires TYPO3 version 4.5.0 or newer)

2012-09-23 Michael Schams <schams.net>

	* TypoScript template uses type=string instead of (undefined) type=input

2012-05-01 Michael Schams <schams.net>

	* documentation updated
	* extension released as version 1.2.4 in TYPO3 Extension Repository (TER)

2012-04-29 Michael Schams <schams.net>

	* PHP function calls require_once replaced with TYPO3 API call: t3lib_div::requireOnce()
	* internal method convertVersionNumberToInteger() implemented (avoids deprecated int_from_ver() but supports old TYPO3 versions)
	* inclusion of file/class t3lib_exec forced for TYPO3 version 4.1.x

2012-02-24 Michael Schams <schams.net>

	* message in FlexForms explicitly warns about the fact that the FE plugin is deprecated
	* message in output explicitly warns about the fact that the FE plugin is deprecated
	* (bugfix) wrong retrieval of values from locallang.xml fixed (FE plugin caused "Call to a member function getLL()..." error)
	* dependency constraints to TYPO3 min/max versions added ("own" ext_emconf.php)

2012-02-19 Michael Schams <schams.net>

	* code for extension list creation restructured
	* extension dependency check implemented (TYPO3 min/max versions)
	* display version of Nagios plugin (passed by GET/POST argument) and IP address of Nagios server implemented (for informational purposes)
	* (bugfix) extension configuration array not initialized correctly
	
2012-01-05 Michael Schams <schams.net>

	* documentation updated
	* extension released as version 1.2.3 in TYPO3 Extension Repository (TER)

2011-12-09 Michael Schams <schams.net>

	* new file/class "tx_nagios_utility_command" replicates t3lib_utility_Command if it does not exist
	* check "diskUsage" implemented (requires UNIX/Linux system)
	* check "sitename" implemented (for informational purposes)
	* check "servername" implemented (for informational purposes)

2011-12-01 Michael Schams <schams.net>

	* documentation updated
	* extension released as version 1.2.2 in TYPO3 Extension Repository (TER)

2011-11-18 Michael Schams <schams.net>

	* backend warning for administrators implemented if list of Nagios servers looks suspecious (requires TYPO3 version 4.2.0 or newer)
	* file name of FlexForms updated
	* unused labels in locallang_tca.xml removed
	* invalid return value "void" removed at several methods to address PHP warnings
	* initialization of extension configuration implemented to address PHP warnings

2011-11-12 Michael Schams <schams.net>

	* file name and class name of FlexForms updated
	* typo in file ext_conf_template.txt corrected

2011-11-07 Michael Schams <schams.net>

	* line after closing php-end-tag removed (impacts generation of XML and other files)

2011-11-05 Michael Schams <schams.net>

	* compatibility with TYPO3 v4.1.x and TYPO3 v4.2.x implemented
	* database checks in eID method fixed
	* isValidNagiosServer() code simplified
	* list of allowed Nagios servers supports wildcards now (e.g. 123.45.67.*)

2011-11-01 Michael Schams <schams.net>

	* corrupted version information fixed

classes/class.tx_nagios_eid.php
<?php
/**
 * This file is part of the TYPO3 CMS Extension "Nagios"
 * Author: Michael Schams - https://schams.net
 *
 * It is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License, either version 2
 * of the License, or any later version.
 *
 * For the full copyright and license information, please read the
 * LICENSE.txt file that was distributed with this source code.
 *
 * The TYPO3 project - inspiring people to share!
 */

/**
 * Class triggered by eID calls
 *
 * @author		Michael Schams <schams.net> and Christian Lerrahn
 * @package		TYPO3
 * @subpackage	tx_nagios
 *
 * public function main()
 *	The main method of the extension.
 *
 * private function getExtensionConfiguration()
 *	Returns extension configuration (via TYPO3 CMS Extension Manager).
 *
 * private function getStaticMarkers()
 *	Returns static markers.
 *
 */

// include class tx_nagios which builds main functions
t3lib_div::requireOnce(t3lib_extMgm::extPath('nagios').'/classes/class.tx_nagios.php');

// compatibility with TYPO3 CMS version 4.1.x and 4.2.x
if(tx_nagios::convertVersionNumberToInteger(TYPO3_version) < 4003000) {
	if(class_exists('tslib_cObj') === FALSE && is_readable(PATH_tslib.'class.tslib_content.php')) {
		t3lib_div::requireOnce(PATH_tslib.'class.tslib_content.php');
	}
	if(class_exists('language') === FALSE && is_readable(PATH_typo3.'sysext/lang/lang.php')) {
		t3lib_div::requireOnce(t3lib_extMgm::extPath('lang', 'lang.php'));
	}
}

class tx_nagios_eid {

	var $scriptRelPath = 'classes/class.tx_nagios_eid.php';
	var $extKey        = 'nagios';

	/**
	 * Details about THIS extension (e.g. title, version, etc.), set in constructor.
	 *
	 * @var array
	 */
	private $extensionDetails = array();

	/**
	 * String (usually a character) used to separate each element.
	 *
	 * @var string
	 */
	private $elementSeparator = "\n";

	/**
	 * This variable enables/disables debug mode.
	 * This setting can be configured in TYPO3 CMS BE (flexforms).
	 *
	 * @var resource
	 */
	private $debugMode = FALSE;

	/**
	 * Extension configuration defined by flexforms in TYPO3 CMS backend
	 *
	 * @var array
	 */
	private $flexFormConfiguration = array();

	/**
	 * Extension configuration defined in Extension Manager (TYPO3 CMS backend)
	 *
	 * @var array
	 */
	private $extensionConfiguration = array();

	/**
	 * Language object (required to access local labels)
	 *
	 * @var object
	 */
	private $language;

	/**
	 * Nagios object (instantiated in Contructor method)
	 *
	 * @var object
	 */
	private $objNagios;

	/**
	 * Local cObj (required for marker substitution)
	 *
	 * @var object
	 */
	private $cObj;


	/**
	 * Constructor method.
	 *
	 * @access	public
	 * @return	void
	 */
	public function __construct() {

		$this->objNagios = t3lib_div::makeInstance('tx_nagios');

		// get details about THIS extension
		$this->extensionDetails = $this->objNagios->getExtensionDetails($this->extKey);
		if($this->extensionDetails === FALSE) {
			$this->extensionDetails = array('title' => 'Nagios', 'version' => 'unknown');
		}

		// create local cObj
		$this->cObj = t3lib_div::makeInstance('tslib_cObj');

		// create language object to retrieve labels (keywords) - TYPO3 CMS version 4.1.x compatible
		$GLOBALS['LANG'] = t3lib_div::makeInstance('language');
		$GLOBALS['LANG']->init('default');
		$GLOBALS['LANG']->includeLLFile('EXT:'.$this->extKey.'/lang/locallang.xml');
		$this->language = $GLOBALS['LANG'];

		// eID method requires a database connect in TYPO3 < 6.1
		if(tx_nagios::convertVersionNumberToInteger(TYPO3_version) < 6001000) {
			tslib_eidtools::connectDB();
		}
	}


	/**
	 * The main method of the extension.
	 *
	 * @access	public
	 * @return	string	Information about the TYPO3 CMS instance according to extension configuration
	 */
	public function main() {

		$this->extensionConfiguration = array_merge($this->objNagios->initExtensionConfiguration(), $this->getExtensionConfiguration());
		$static_markers = $this->getStaticMarkers();

		$data = array();

		// get version of Nagios plugin (client) if passed by GET/POST argument
		$nagiosPluginVersion = $this->objNagios->getNagiosPluginVersion();

		if($this->extensionConfiguration['featureDebugMode'] == 1) {
			$this->debugMode = TRUE;
		}

		// do *not* take proxy servers into account by default when checking client access, but if
		// extension security exception has been explicitly confirmed (securityProxyHeaders = 1),
		// HTTP headers typically passed through by proxy servers are investigated in addition to
		// $_SERVER['REMOTE_ADDR']
		if($this->extensionConfiguration['securityProxyHeaders'] != 1) {
			$this->objNagios->takeProxyServerIntoAccount(FALSE);
		}

		if($this->objNagios->isValidNagiosServer($this->extensionConfiguration['securityNagiosServerList']) === FALSE) {

			// remote (Nagios?) server is not allowed to retrieve information about this TYPO3 CMS server.
			// generate some further explanations (problem description) as comments.
			$data[] = '# '.$this->language->getLL('access_denied_line01');
			$data[] = '# '.$this->language->getLL('access_denied_line02');
			$data[] = '# '.$this->language->getLL('access_denied_line03');
			$data[] = '# (extension configuration via TYPO3 CMS Extension Manager)';
			$data[] = $this->language->getLL('keyword_message').':access denied';
		}
		else {
			// all security checks successfully passed - Nagios server is allowed to retrieve data

			if($this->extensionConfiguration['featureTYPO3Version'] != 0) {
				$data[] = $this->language->getLL('keyword_typo3').':version-'.$this->objNagios->getTypo3Version();
			}
			if($this->extensionConfiguration['featurePHPVersion'] != 0) {
				$data[] = $this->language->getLL('keyword_php').':version-'.$this->objNagios->getPhpVersion();
			}

			// 'featureLoadedExtensionsOnly' must be set *prior* $this->objNagios->getExtensionVersions()
			if($this->extensionConfiguration['featureLoadedExtensionsOnly'] == 1) {
				$this->objNagios->setLoadedExtensionsOnly(TRUE);
			}
			if($this->extensionConfiguration['featureExtensionList'] != 0) {
				$data[] = implode($this->elementSeparator, $this->objNagios->getExtensionVersions($this->language->getLL('keyword_extension')));
			}
			if($this->extensionConfiguration['featureExtensionDependencyTypo3'] != 0) {
				$extensionDependencyIssues = $this->objNagios->getExtensionDependency($this->language->getLL('keyword_extensiondependencytypo3'), tx_nagios::TX_NAGIOS_EXTENSION_DEPENDENCY_TYPO3);
				if(is_array($extensionDependencyIssues) && count($extensionDependencyIssues) > 0) {
					$data[] = implode($this->elementSeparator, $extensionDependencyIssues);
				}
			}
			if($this->extensionConfiguration['featureSitename'] != 0) {
				$data[] = $this->language->getLL('keyword_sitename').':'.$this->objNagios->getSitename();
			}
			if($this->extensionConfiguration['featureServername'] != 0) {
				$data[] = $this->language->getLL('keyword_servername').':'.$this->objNagios->getServername();
			}
			if($this->extensionConfiguration['featureBasicDatabaseDetailsTables'] != 0) {
				$data[] = $this->language->getLL('keyword_databasetables').':'.$this->objNagios->getDatabaseDetailsTables();
			}
			if($this->extensionConfiguration['featureBasicDatabaseDetailsProcesslist'] != 0) {
				$database_processlist = $this->objNagios->getDatabaseDetailsProcesses();
				if($database_processlist !== FALSE && is_numeric($database_processlist)) {
					$data[] = $this->language->getLL('keyword_databaseprocesslist').':'.$database_processlist;
				}
			}
			if($this->extensionConfiguration['featureTimestamp'] != 0) {
				$data[] = $this->language->getLL('keyword_timestamp').':'.date('U-T');
			}
			if($this->extensionConfiguration['featureDeprecationLog'] != 0) {
				$data[] = $this->language->getLL('keyword_deprecationlog').':'.$this->objNagios->getDeprecationLogSetting();
			}
			if($this->extensionConfiguration['featureDonationNotice'] != 0) {
				$data[] = $this->language->getLL('keyword_donationnotice').':'.$this->objNagios->getDonationNoticeSetting();
			}
			if($this->extensionConfiguration['featureCheckDiskUsage'] != 0) {
				$data[] = $this->language->getLL('keyword_diskusage').':'.$this->objNagios->getDiskUsage();
			}
		}

		// glue single elements together, using $this->elementSeparator variable
		$data = implode($this->elementSeparator, $data);

		// add header ("Nagios TYPO3 Monitoring Version x.x.x...") if not permitted in extension configuration
		$header = '';
		if($this->extensionConfiguration['securitySupressHeader'] != 1) {
			$header.= '# '.$this->extensionDetails['title'].' Version '.$this->extensionDetails['version'].' - https://schams.net/nagios' . PHP_EOL;

			if($nagiosPluginVersion !== FALSE && is_string($nagiosPluginVersion) && !empty($nagiosPluginVersion)) {
				$header.= '# Nagios Plugin Version '.$nagiosPluginVersion.' (IP: ' . t3lib_div::getIndpEnv('REMOTE_ADDR') . ')' . PHP_EOL;
			}
			$header.= "\n";
		}

//		if(!headers_sent() && error_get_last() == NULL) {
		if(!headers_sent()) {

			// force text/plain in HTTP header
			header('Content-Type: text/plain');

			// disable client side caching, e.g. proxy servers (very basic only)
			header('Cache-Control: no-cache, must-revalidate');
			header('Expires: Sat, 01 Jan 2011 00:00:00 UTC');
		}

		// final output: header (if available) + data (with static markers replaced)
		echo $header.$this->cObj->substituteMarkerArray($data, $static_markers);
	}


	/**
	 * Returns extension configuration (via TYPO3 CMS Extension Manager).
	 *
	 * @access	private
	 * @return	array	Extension configuration (key: keyword, value: value)
	 */
	private function getExtensionConfiguration() {

		if(isset($GLOBALS['TYPO3_CONF_VARS']['EXT']['extConf'][$this->extKey])) {
			return unserialize($GLOBALS['TYPO3_CONF_VARS']['EXT']['extConf'][$this->extKey]);
		}
		return array();
	}


	/**
	 * Returns static markers.
	 *
	 * @access	private
	 * @return	array	Array with static markers (key: keyword, value: value)
	 */
	private function getStaticMarkers() {

		return array(
		);
	}
}

$nagiosInfo = t3lib_div::makeInstance('tx_nagios_eid');
$nagiosInfo->main();
